File: Standalone/Web/NowUI.Web/wwwroot/nowui/results.js
// W5 - the result table, JavaScript half. Docs/Standalone/M3-Spec.md section 6.
//
// This is the answer half of the bridge: what a JavaScript author reads synchronously when they write
//
//     if (ui.button('Add')) addPerson();
//     state.name = ui.textField('name', state.name);
//
// and it is one frame old (section 6.2). The table it reads was written by the LAST replay, rode back inside the
// same `record` call this frame's ops go out on, and describes controls that were drawn then.
//
// Four behaviours live here, and three of the four are the kind that fail silently:
//
//   1. THE GENERATION STAMP. A record is "present" only if it was written for the frame currently being loaded.
//      `results[rid]` is an array index and a comparison - no clearing pass, no allocation - and a control that
//      was NOT drawn last frame simply has a stale stamp, so it reads as absent (section 6.1). That is what stops
//      the table serving a `clicked` for a button that stopped existing three seconds ago.
//
//   2. ONE-SHOT LATCHING (section 6.3). `clicked`, `submitted`, `dragStarted`, `dragEnded` and `cancelled` are
//      events, not states. The loader marks them unread; the first read returns true and clears the bit. So a
//      control read twice in one frame fires once, and an event can never be delivered late, because the record
//      is replaced wholesale on the next load.
//
//   3. THE VALUE RECONCILIATION RULE (section 6.4). The rule that is worth reading the file for; it has its own
//      comment on `value` below, because the plausible version of it drops a keystroke.
//
//   4. THE MISSED READ (section 6.4, last paragraph). A read for a rid with no record returns the identity
//      default - false for an event, the caller's own value for a value - and under `?nowui=debug` a control that
//      has been drawn for three consecutive frames without ever producing a record is reported once. That is the
//      signature of a control the replay is silently not reaching, and it is otherwise indistinguishable from a
//      button nobody clicked.
//
// Nothing here touches the DOM or the wasm runtime, so Standalone/NowUI.Bridge.Tests/js drives it under node.

// ------------------------------------------------------------------------------------------------ the wire
//
// The twin of Standalone/NowUI.Bridge/Results.cs. Both halves have to agree about the record layout, and unlike
// the op stream there is no surface hash over it - the table is written and read by two files that ship together
// in the same wasm build, and a mismatch shows up as a decoded record with a nonsense kind, which `load` refuses.

/// Section 6.1's header, in front of the first record. Two slots, and Results.cs says why they exist.
export const HDR_RECORD_COUNT = 0;
export const HDR_TEXT_BYTES = 1;
export const HDR_SLOTS = 2;

/// Section 6.1's value kinds.
export const KIND_NONE = 0;
export const KIND_F32 = 1;
export const KIND_I32 = 2;
export const KIND_BOOL = 3;
export const KIND_STR = 4;
export const KIND_COLOR = 5;

/// W6. Two slots, low word first. ui.datePicker's value is a Unix epoch in milliseconds; an f32 carries 24 bits
/// of mantissa, so every date in the picker's range would land on the same handful of representable days.
export const KIND_I64 = 6;

/// Section 6.1's flags, as they sit in the high 24 bits of the record's second slot.
export const F_CLICKED = 1 << 0;
export const F_SUBMITTED = 1 << 1;
export const F_DRAG_STARTED = 1 << 2;
export const F_DRAG_ENDED = 1 << 3;
export const F_CANCELLED = 1 << 4;
export const F_CHANGED = 1 << 5;
export const F_FOCUSED = 1 << 6;
export const F_HOVERED = 1 << 7;
export const F_PRESSED = 1 << 8;
export const F_HELD = 1 << 9;
export const F_RELEASED = 1 << 10;
export const F_DRAGGING = 1 << 11;
export const F_HAS_RECT = 1 << 12;
export const F_PRESENT = 1 << 13;

/// How far the flag field is shifted up in that slot; the low 8 bits are the value kind.
export const FLAGS_SHIFT = 8;
export const KIND_MASK = 0xff;

/// Section 6.3: the flags that are events rather than states, and are therefore consumed by the first read.
export const ONE_SHOT = F_CLICKED | F_SUBMITTED | F_DRAG_STARTED | F_DRAG_ENDED | F_CANCELLED;

/// Section 6.4: how many consecutive drawn-but-never-recorded frames before a rid is reported under debug.
const MISS_REPORT_AT = 3;

const decoder = new TextDecoder();

export class Results {
    constructor(options = {}) {
        this.frame = -1;
        this.debug = options.debug === true;
        this.onReport = options.onReport || defaultReport;

        // rid-indexed, and grown together. Parallel typed arrays rather than an object per rid: a record is read
        // on the hot path of every control in the frame, and `this.stamp[rid] === this.frame` is the whole of the
        // presence test.
        this._grow(64);

        // Section 6.4's memory: what this side RETURNED for each rid last frame. Not what the control produced,
        // and not what the author passed - what the author was told. See `value`.
        this.lastSent = [];

        // Diagnostics. `records` is the last load's size, and `reports` is what the missed-read check said.
        this.records = 0;
        this.reports = [];
    }

    // -------------------------------------------------------------------------------------------- loading

    /// Reads one frame's table. `slots` is the Int32Array `record` sliced out of the managed results view (or null
    /// before the first replay has run), `text` the UTF-8 bytes its string values point into, `frame` the frame
    /// number that is about to be drawn.
    ///
    /// Called exactly once per frame, BEFORE the author's draw function - and not at all on a NEED_MORE retry,
    /// because a retry re-uses the bytes already recorded and must not consume the one-shot events a second time.
    load(slots, text, frame) {
        this.frame = frame;
        this.records = 0;

        if (!slots || slots.length < HDR_SLOTS) return;   // no table yet: the first frame of the session

        const count = slots[HDR_RECORD_COUNT];
        if (count <= 0) return;

        // The f32 view over the same bytes. `slice()` hands back a copy with its own buffer, so this is a view
        // over JavaScript's own memory and stays valid for as long as the load does.
        const f32 = new Float32Array(slots.buffer, slots.byteOffset, slots.length);

        let i = HDR_SLOTS;

        for (let n = 0; n < count; n++) {
            if (i + 2 > slots.length) {
                // A truncated table. Everything read so far is good and is kept; the rest is dropped rather than
                // decoded out of whatever follows it in the buffer.
                this._report('NowUI: the result table declared ' + count + ' records and ran out of slots after ' +
                    n + '. This is a bridge bug.');
                break;
            }

            const rid = slots[i++];
            const head = slots[i++];
            const kind = head & KIND_MASK;
            const flags = head >>> FLAGS_SHIFT;

            if (rid < 0 || kind > KIND_I64) {
                this._report('NowUI: the result table holds a record with rid ' + rid + ' and value kind ' + kind +
                    ', which this build does not know. This is a bridge bug.');
                break;
            }

            if (rid >= this.stamp.length) this._grow(rid + 1);

            this.stamp[rid] = frame;
            this.flags[rid] = flags;
            this.unread[rid] = flags & ONE_SHOT;
            this.kind[rid] = kind;
            this.num[rid] = 0;
            this.str[rid] = null;
            this.rect[rid] = null;
            this.seen[rid] = 1;
            this.missed[rid] = 0;

            if (kind === KIND_STR) {
                const offset = slots[i++];
                const length = slots[i++];
                this.str[rid] = length > 0 && text ? decoder.decode(text.subarray(offset, offset + length)) : '';
            } else if (kind === KIND_F32) {
                this.num[rid] = f32[i++];
            } else if (kind === KIND_I32) {
                this.num[rid] = slots[i++];
            } else if (kind === KIND_BOOL) {
                this.num[rid] = slots[i++] !== 0 ? 1 : 0;
            } else if (kind === KIND_COLOR) {
                this.num[rid] = slots[i++] >>> 0;
            } else if (kind === KIND_I64) {
                // Reassembled into a double rather than a BigInt: the values this kind carries are epoch
                // milliseconds, which stay exact to 2^53 - a quarter of a million years either side of 1970 -
                // and a BigInt would make `state.when = ui.datePicker(...)` return something an author cannot
                // hand to `new Date()` without a cast.
                const low = slots[i] >>> 0;
                const high = slots[i + 1];
                i += 2;
                this.num[rid] = high * 4294967296 + low;
            }

            if ((flags & F_HAS_RECT) !== 0) {
                this.rect[rid] = [f32[i], f32[i + 1], f32[i + 2], f32[i + 3]];
                i += 4;
            }

            this.records++;
        }
    }

    // -------------------------------------------------------------------------------------------- reading

    /// True when this rid's record was written for the frame being drawn.
    present(rid) {
        return rid < this.stamp.length && this.stamp[rid] === this.frame;
    }

    /// Section 6.3. Reads a one-shot event and consumes it, so a second read this frame is false.
    ///
    /// Every action control in the surface goes through here, which is why the identity default is in one place:
    /// a rid with no record returns false. A button that has never been drawn, a button drawn for the first time
    /// this frame, and a button nobody clicked are all indistinguishable here - and all three should be false.
    event(rid, bit) {
        if (!this.present(rid)) { this._miss(rid); return false; }

        const hit = (this.unread[rid] & bit) !== 0;
        if (hit) this.unread[rid] &= ~bit;
        return hit;
    }

    /// A state flag - hovered, focused, changed. Read as often as the author likes; never consumed.
    state(rid, bit) {
        if (!this.present(rid)) { this._miss(rid); return false; }
        return (this.flags[rid] & bit) !== 0;
    }

    /// The post-draw value the last replay recorded for this rid, without the reconciliation of `value`. Used by
    /// the handler prologue in nowui.js, which has to hand a submitted field's text to onSubmit BEFORE the author's
    /// draw function runs and therefore has no incoming value to reconcile against.
    current(rid) {
        return this.present(rid) ? this._read(rid) : undefined;
    }

    /// The rect of a control drawn with a rect in its record, or null. Section 6.1: rects are opt-in.
    rectOf(rid) {
        return this.present(rid) ? this.rect[rid] : null;
    }

    // ------------------------------------------------------------------------------------ section 6.4

    /// THE VALUE RECONCILIATION RULE. Every value control's return goes through here, and the return value is
    /// ALSO what the caller must emit - resolving before emitting is half of what makes this correct.
    ///
    /// The rule: the recorder remembers what it RETURNED for each rid last frame. If the incoming value differs,
    /// the author wrote it and the author wins. If it is identical, the author is echoing, and the freshest known
    /// value wins.
    ///
    /// Why not the obvious version. Design B emitted the caller's value and then consulted the result:
    ///
    ///     emit(OP_TEXT_FIELD, rid, value);          // WRONG
    ///     const r = results.get(rid);
    ///     return (r && r.changed) ? r.value : value;
    ///
    /// Trace a keystroke. Frame N-1's replay produced "abc" from "ab" and set `changed`. At frame N the author's
    /// state.name is still "ab" - it only receives "abc" from this call's RETURN. So the op carries "ab", and
    /// frame N's replay calls Draw(ref text) with "ab". The caller's string is authoritative: NowTextField stores
    /// no text of its own and clamps its edit state to whatever it is handed (NowTextField.cs:1398,
    /// NowTextEdit.Clamp(ref state, text)). The character is reverted, one frame after it was typed.
    ///
    /// Resolving before emitting fixes that and introduces the opposite bug on its own: a programmatic write -
    /// `state.name = ''` in a Clear handler - would be overwritten by the last frame's result forever. The
    /// `lastSent` comparison is what tells the two apart, and it is a comparison against what this side returned,
    /// not against what the control produced.
    ///
    /// Four properties, each of which the two-line version lacks:
    ///   * a keystroke is never reverted, because the emitted value is the resolved one;
    ///   * a programmatic write always lands;
    ///   * the author never observes a stale value - their own, or a fresher one;
    ///   * it is bit-identical when nothing changed: no drift, no snapping, no re-clamping.
    value(rid, incoming) {
        if (rid >= this.stamp.length) this._grow(rid + 1);

        const prev = this.lastSent[rid];
        let v = incoming;

        if (prev !== undefined && Object.is(incoming, prev)) {
            if (this.present(rid)) v = this._read(rid);
            else this._miss(rid);
        }

        this.lastSent[rid] = v;
        return v;
    }

    /// The post-draw value the last replay recorded for this rid, in the kind the record declared.
    _read(rid) {
        switch (this.kind[rid]) {
            case KIND_STR: return this.str[rid];
            case KIND_BOOL: return this.num[rid] !== 0;
            case KIND_F32:
            case KIND_I32:
            case KIND_COLOR:
            case KIND_I64: return this.num[rid];
            default: return undefined;
        }
    }

    // ---------------------------------------------------------------------------------- the missed read

    /// Section 6.4's last paragraph. Counts reads that could not be served and reports the shape that matters:
    /// a control drawn for three consecutive frames that has NEVER produced a record. A control drawn for the
    /// first time, and one whose key just changed, both miss once and then start being served - so neither is
    /// reported, which is the whole reason the check is "three consecutive AND never seen" rather than "missed".
    _miss(rid) {
        if (!this.debug || rid >= this.stamp.length) return;

        this.missed[rid]++;

        if (this.missed[rid] >= MISS_REPORT_AT && this.seen[rid] === 0 && this.reported[rid] === 0) {
            this.reported[rid] = 1;
            this._report(
                'NowUI: rid ' + rid + ' has been drawn for ' + this.missed[rid] + ' consecutive frames and the ' +
                'replay has never produced a result for it. Every read of it returns the identity default - false ' +
                'for an event, your own value for a value - so this control looks like one nobody has touched.\n' +
                'That is the signature of a control the replay is not reaching: an op with no case in the decoder, ' +
                'or a control drawn inside a scope the replay closed early.');
        }
    }

    _report(message) {
        this.reports.push(message);
        this.onReport(message);
    }

    // -------------------------------------------------------------------------------------------- growth

    _grow(need) {
        let size = this.stamp === undefined ? 64 : this.stamp.length;
        while (size < need) size *= 2;

        const stamp = new Int32Array(size).fill(-1);
        const flags = new Int32Array(size);
        const unread = new Int32Array(size);
        const kind = new Int32Array(size);
        const num = new Float64Array(size);
        const seen = new Uint8Array(size);
        const missed = new Int32Array(size);
        const reported = new Uint8Array(size);

        if (this.stamp !== undefined) {
            stamp.set(this.stamp);
            flags.set(this.flags);
            unread.set(this.unread);
            kind.set(this.kind);
            num.set(this.num);
            seen.set(this.seen);
            missed.set(this.missed);
            reported.set(this.reported);
        }

        this.stamp = stamp;
        this.flags = flags;
        this.unread = unread;
        this.kind = kind;
        this.num = num;
        this.seen = seen;
        this.missed = missed;
        this.reported = reported;

        if (this.str === undefined) { this.str = []; this.rect = []; }
    }
}

function defaultReport(message) {
    if (typeof console !== 'undefined' && console.warn) console.warn(message);
}

File: Standalone/NowUI.Bridge/Results.cs
// W5 - the result table, managed half. Docs/Standalone/M3-Spec.md section 6.
//
// One record per interactive control drawn in the last REAL replay pass, in draw order, written into a buffer that
// rides back to JavaScript inside the same `record` call the ops go out on (section 6.1) - so the whole answer half
// of the bridge costs no boundary crossing of its own.
//
// Three properties this file exists to hold, each of which is silent when it breaks:
//
//   1. MEASURE-PASS SUPPRESSION (section 5.6). Under exactLayout the buffer is decoded twice and the FIRST pass is
//      passive: NowLayout.isMeasurePass is true, NowInput.isPassive is true, and every interaction result the
//      controls hand back is inert. A writer that recorded that pass would overwrite real results with zeroes,
//      and it would do it on the pass that runs FIRST, so the damage would look like "the click was never seen".
//      Write() refuses while either flag is set, and counts what it refused, so the suppression is a measured
//      number in a test rather than a claim in a comment.
//
//   2. CONTROLS NOT DRAWN PRODUCE NO RECORD (section 6.1). The cursor is reset at the start of every pass and the
//      table is sealed at the end of the frame, so the table is exactly what this frame drew - never a merge with
//      what the last one did. That is what stops a `clicked` being delivered for a button that stopped existing.
//
//   3. THE HEADER. Section 6.1's layout starts at the first record; this adds two slots in front of it, holding
//      the record count and the byte count of the string section. The reason is the ABI, not the table: `record`
//      takes ONE resultSlots argument (section 5.1) and the strings live in a second view whose used length would
//      otherwise be unknowable to JavaScript - it would have to slice() the whole text buffer every frame, which
//      grows without bound as soon as one text field holds a long string. Two slots buy that back. The records
//      themselves are section 6.1's, unchanged, and start at HeaderSlots.
//
// What this file does NOT do is decide which flags a control reports. That is the decoder's job, because it is the
// decoder that holds the value the control's consumer returned - and for several tier-1 controls that value is one
// bool. See BridgeReplay for the honest note about what NowButton.Draw() can and cannot tell the table.

using System;
using System.Text;
using NowUI;

namespace NowUI.Bridge
{
    /// <summary>Section 6.1's value kinds, with the number of slots each occupies after the two-slot record head.</summary>
    public enum BridgeValueKind
    {
        /// <summary>No value slots. An action control: the flags are the whole answer.</summary>
        None = 0,

        /// <summary>One slot, read through the Float32 view.</summary>
        F32 = 1,

        /// <summary>One slot.</summary>
        I32 = 2,

        /// <summary>One slot, 0 or 1.</summary>
        Bool = 3,

        /// <summary>Two slots: byteOffset and byteLength into the results text buffer.</summary>
        Str = 4,

        /// <summary>One slot, RGBA8 packed.</summary>
        Color = 5,

        /// <summary>
        /// W6. Two slots, low word first. <c>ui.datePicker</c>'s value is a Unix epoch in milliseconds and does
        /// not survive an f32 - 2026 is about 1.77e12, and a float carries 24 bits of mantissa, so the nearest
        /// representable neighbours are a quarter of a million milliseconds apart. Every date in the picker's
        /// range would snap to the same handful of days.
        /// </summary>
        I64 = 6,
    }

    /// <summary>
    /// Section 6.1's flag set. Twenty-four bits are available and fourteen are named; the rest are reserved so a
    /// later flag does not renumber one JavaScript already reads.
    /// </summary>
    [Flags]
    public enum BridgeFlags
    {
        None = 0,

        // --- one-shot events (section 6.3): the loader marks them unread and the first read clears them --------
        Clicked = 1 << 0,
        Submitted = 1 << 1,
        DragStarted = 1 << 2,
        DragEnded = 1 << 3,
        Cancelled = 1 << 4,

        // --- states -------------------------------------------------------------------------------------------
        Changed = 1 << 5,
        Focused = 1 << 6,
        Hovered = 1 << 7,
        Pressed = 1 << 8,
        Held = 1 << 9,
        Released = 1 << 10,
        Dragging = 1 << 11,

        // --- structure ----------------------------------------------------------------------------------------
        /// <summary>Four slots of rect follow the value slots.</summary>
        HasRect = 1 << 12,

        /// <summary>The control was drawn. Always set on a written record; it is what a reader tests.</summary>
        Present = 1 << 13,
    }

    /// <summary>The result table: record layout, the write path, and measure-pass suppression.</summary>
    public sealed class BridgeResults
    {
        /// <summary>Slot 0 of the buffer: how many records follow the header.</summary>
        public const int HdrRecordCount = 0;

        /// <summary>Slot 1: how many bytes of the results text buffer this frame used.</summary>
        public const int HdrTextBytes = 1;

        /// <summary>The number of header slots in front of the first record.</summary>
        public const int HeaderSlots = 2;

        /// <summary>Section 6.1's second slot: flags in the high 24 bits, valueKind in the low 8.</summary>
        public const int ValueKindMask = 0xFF;

        /// <summary>How far the flag field is shifted up in the header slot.</summary>
        public const int FlagsShift = 8;

        // Sized so the 23 controls of section 1 (about 100 slots, section 6.1) fit without a growth.
        private const int InitialSlots = 256;
        private const int InitialTextBytes = 1024;

        private int[] m_Slots = new int[InitialSlots];
        private byte[] m_Text = new byte[InitialTextBytes];

        private int m_Cursor;
        private int m_TextCursor;
        private int m_Records;
        private int m_Sealed;
        private int m_Suppressed;

        /// <summary>The slot buffer, for the span handed to <c>record</c>. Re-read every frame: it grows.</summary>
        public int[] slots => m_Slots;

        /// <summary>The string bytes every <see cref="BridgeValueKind.Str"/> value points into.</summary>
        public byte[] text => m_Text;

        /// <summary>
        /// How many slots of <see cref="slots"/> the last sealed frame filled, header included. This is the
        /// <c>resultSlots</c> argument of section 5.1's <c>record</c>, and it is zero before the first frame has
        /// been replayed - which JavaScript reads as "no table", not as an error.
        /// </summary>
        public int sealedSlots => m_Sealed;

        /// <summary>How many records the last sealed frame wrote.</summary>
        public int recordCount { get; private set; }

        /// <summary>
        /// How many writes were refused because the pass was passive. Under <c>exactLayout</c> this is one per
        /// record per frame and under a single-pass frame it is zero; a test asserts exactly that, which is what
        /// makes "a result written during a measure pass never reaches JavaScript" a measurement.
        /// </summary>
        public int suppressedWrites => m_Suppressed;

        /// <summary>
        /// True while NowUI is measuring rather than drawing. Both halves matter and neither implies the other:
        /// <c>isMeasurePass</c> is NowLayout's own flag (NowLayout.cs:1256) and <c>isPassive</c> is the input
        /// system's (NowInput.cs:1317), and a control asked to draw while either is set reports nothing real.
        /// </summary>
        public static bool passive => NowLayout.isMeasurePass || NowInput.isPassive;

        // ------------------------------------------------------------------------------------------- the frame

        /// <summary>Starts a frame. Resets the counters the whole frame accumulates.</summary>
        public void BeginFrame()
        {
            m_Suppressed = 0;
            BeginPass();
        }

        /// <summary>
        /// Starts one replay pass. Resets the write cursor, so a second pass overwrites the first rather than
        /// appending to it - the property that makes the table "what this frame drew" rather than a merge.
        /// </summary>
        public void BeginPass()
        {
            m_Cursor = HeaderSlots;
            m_TextCursor = 0;
            m_Records = 0;
        }

        /// <summary>
        /// Closes the frame and writes the header. Called after the last pass; nothing crosses the boundary until
        /// it has run, so a frame that threw mid-replay hands JavaScript the previous table rather than half of a
        /// new one.
        /// </summary>
        public void Seal()
        {
            m_Slots[HdrRecordCount] = m_Records;
            m_Slots[HdrTextBytes] = m_TextCursor;
            m_Sealed = m_Cursor;
            recordCount = m_Records;
        }

        // ------------------------------------------------------------------------------------------- writing

        /// <summary>An action control: flags only, no value (section 6.1's <c>None</c> kind).</summary>
        public void WriteEvent(int rid, BridgeFlags flags)
        {
            Write(rid, flags, BridgeValueKind.None, 0, 0, null);
        }

        /// <summary>A boolean-valued control.</summary>
        public void WriteBool(int rid, BridgeFlags flags, bool value)
        {
            Write(rid, flags, BridgeValueKind.Bool, value ? 1 : 0, 0, null);
        }

        /// <summary>A float-valued control. The value is stored bit-for-bit and read through the Float32 view.</summary>
        public void WriteF32(int rid, BridgeFlags flags, float value)
        {
            Write(rid, flags, BridgeValueKind.F32, BitConverter.SingleToInt32Bits(value), 0, null);
        }

        /// <summary>An int-valued control.</summary>
        public void WriteI32(int rid, BridgeFlags flags, int value)
        {
            Write(rid, flags, BridgeValueKind.I32, value, 0, null);
        }

        /// <summary>A 64-bit-valued control: two slots, low word first. See <see cref="BridgeValueKind.I64"/>.</summary>
        public void WriteI64(int rid, BridgeFlags flags, long value)
        {
            Write(rid, flags, BridgeValueKind.I64, unchecked((int)(value & 0xFFFFFFFF)), (int)(value >> 32), null);
        }

        /// <summary>
        /// A string-valued control. The bytes go into the text buffer and the record carries (offset, length),
        /// which is the same shape the op stream's volatile table uses in the other direction (section 5.4).
        /// </summary>
        public void WriteString(int rid, BridgeFlags flags, string value)
        {
            Write(rid, flags, BridgeValueKind.Str, 0, 0, value ?? string.Empty);
        }

        /// <summary>Adds the four rect slots to the record just written, and sets <see cref="BridgeFlags.HasRect"/>.</summary>
        /// <remarks>
        /// Called immediately after a Write, and only then. Rects are opt-in (section 6.1) because
        /// <c>bool Draw()</c> discards them; a consumer that returns a result struct hands one back for free, and
        /// this is how it reaches the table without a second call shape.
        /// </remarks>
        public void AppendRect(NowRect rect)
        {
            if (passive || m_Records == 0) return;

            // The record head is at the slot the last Write started at; find it by walking back is not possible,
            // so the flag is set on the head slot remembered by Write.
            m_Slots[m_LastHead] |= ((int)BridgeFlags.HasRect) << FlagsShift;

            EnsureSlots(m_Cursor + 4);
            m_Slots[m_Cursor++] = BitConverter.SingleToInt32Bits(rect.x);
            m_Slots[m_Cursor++] = BitConverter.SingleToInt32Bits(rect.y);
            m_Slots[m_Cursor++] = BitConverter.SingleToInt32Bits(rect.width);
            m_Slots[m_Cursor++] = BitConverter.SingleToInt32Bits(rect.height);
        }

        private int m_LastHead;

        private void Write(int rid, BridgeFlags flags, BridgeValueKind kind, int a, int b, string str)
        {
            // Property 1. This is the whole of measure-pass suppression, and it is one branch because it has to be
            // on the hot path of every control in the frame.
            if (passive)
            {
                ++m_Suppressed;
                return;
            }

            int valueSlots = ValueSlots(kind);
            EnsureSlots(m_Cursor + 2 + valueSlots);

            m_Slots[m_Cursor++] = rid;
            m_LastHead = m_Cursor;
            m_Slots[m_Cursor++] = (((int)(flags | BridgeFlags.Present)) << FlagsShift) | ((int)kind & ValueKindMask);

            switch (kind)
            {
                case BridgeValueKind.None:
                    break;

                case BridgeValueKind.Str:
                {
                    int offset = m_TextCursor;
                    int length = WriteText(str);
                    m_Slots[m_Cursor++] = offset;
                    m_Slots[m_Cursor++] = length;
                    break;
                }

                case BridgeValueKind.I64:
                    m_Slots[m_Cursor++] = a;
                    m_Slots[m_Cursor++] = b;
                    break;

                default:
                    m_Slots[m_Cursor++] = a;
                    break;
            }

            ++m_Records;
        }

        /// <summary>How many value slots a record of this kind carries after its two-slot head.</summary>
        private static int ValueSlots(BridgeValueKind kind)
        {
            switch (kind)
            {
                case BridgeValueKind.None: return 0;
                case BridgeValueKind.Str: return 2;
                case BridgeValueKind.I64: return 2;
                default: return 1;
            }
        }

        private int WriteText(string value)
        {
            if (string.IsNullOrEmpty(value)) return 0;

            int maxBytes = Encoding.UTF8.GetMaxByteCount(value.Length);
            EnsureText(m_TextCursor + maxBytes);

            int written = Encoding.UTF8.GetBytes(value, 0, value.Length, m_Text, m_TextCursor);
            m_TextCursor += written;
            return written;
        }

        // ------------------------------------------------------------------------------------------- growth
        //
        // Growth here needs no protocol, unlike the op stream's (section 5.1). The managed side owns both buffers
        // and writes them BETWEEN crossings - the table is written during frame N's replay and read at frame
        // N+1's record - so a grow is an array copy and the next crossing simply carries a longer span.

        private void EnsureSlots(int need)
        {
            if (need <= m_Slots.Length) return;
            int size = m_Slots.Length;
            while (size < need) size *= 2;
            var grown = new int[size];
            Array.Copy(m_Slots, grown, m_Cursor);
            m_Slots = grown;
        }

        private void EnsureText(int need)
        {
            if (need <= m_Text.Length) return;
            int size = m_Text.Length;
            while (size < need) size *= 2;
            var grown = new byte[size];
            Array.Copy(m_Text, grown, m_TextCursor);
            m_Text = grown;
        }

        // ------------------------------------------------------------------------------------------- reading
        //
        // The bridge itself never reads the table - JavaScript does. These exist so a test can assert what was
        // written without re-implementing the walk, and so a diagnostic can print it.

        /// <summary>One decoded record. What JavaScript's loader builds, in C#, for tests and diagnostics.</summary>
        public readonly struct Record
        {
            internal Record(int rid, BridgeFlags flags, BridgeValueKind kind, int raw, string text, NowRect rect, bool hasRect, long wide = 0)
            {
                this.wide = wide;
                this.rid = rid;
                this.flags = flags;
                this.kind = kind;
                this.raw = raw;
                this.text = text;
                this.rect = rect;
                this.hasRect = hasRect;
            }

            public int rid { get; }
            public BridgeFlags flags { get; }
            public BridgeValueKind kind { get; }

            /// <summary>The value slot, unconverted. Use <see cref="f32"/> / <see cref="boolean"/> for the rest.</summary>
            public int raw { get; }

            public string text { get; }
            public NowRect rect { get; }
            public bool hasRect { get; }

            /// <summary>The 64-bit value of an <see cref="BridgeValueKind.I64"/> record; zero otherwise.</summary>
            public long wide { get; }

            public float f32 => BitConverter.Int32BitsToSingle(raw);
            public bool boolean => raw != 0;
            public bool Has(BridgeFlags flag) => (flags & flag) != 0;
        }

        /// <summary>Walks the sealed table. Empty before the first frame has been replayed.</summary>
        public Record[] Decode()
        {
            if (m_Sealed < HeaderSlots) return Array.Empty<Record>();

            int count = m_Slots[HdrRecordCount];
            var records = new Record[count];
            int slot = HeaderSlots;

            for (int i = 0; i < count; ++i)
            {
                int rid = m_Slots[slot++];
                int head = m_Slots[slot++];
                var kind = (BridgeValueKind)(head & ValueKindMask);
                var flags = (BridgeFlags)(head >> FlagsShift);

                int raw = 0;
                string str = null;
                long wide = 0;

                if (kind == BridgeValueKind.Str)
                {
                    int offset = m_Slots[slot++];
                    int length = m_Slots[slot++];
                    str = Encoding.UTF8.GetString(m_Text, offset, length);
                }
                else if (kind == BridgeValueKind.I64)
                {
                    uint low = unchecked((uint)m_Slots[slot++]);
                    int high = m_Slots[slot++];
                    wide = unchecked(((long)high << 32) | low);
                    raw = unchecked((int)low);
                }
                else if (kind != BridgeValueKind.None)
                {
                    raw = m_Slots[slot++];
                }

                NowRect rect = default;
                bool hasRect = (flags & BridgeFlags.HasRect) != 0;
                if (hasRect)
                {
                    rect = new NowRect(
                        BitConverter.Int32BitsToSingle(m_Slots[slot]),
                        BitConverter.Int32BitsToSingle(m_Slots[slot + 1]),
                        BitConverter.Int32BitsToSingle(m_Slots[slot + 2]),
                        BitConverter.Int32BitsToSingle(m_Slots[slot + 3]));
                    slot += 4;
                }

                records[i] = new Record(rid, flags, kind, raw, str, rect, hasRect, wide);
            }

            return records;
        }
    }
}
